### PR TITLE
make: Support 'piecemeal' action modifier

### LINKED ISSUE #42: Support `piecemeal` action modifier
Support `actions piecemeal`, which will split the source list into multiple invocations to fit in the systems command buffer.

- [ ] Create test
- [ ] Support in `Processor::_MakeTarget`

### PR DESCRIPTION
**Pull request by dominicm00**
Closes #42 

- [x] Add piecemeal tests
- [x] Add piecemeal functionality
- [x] Create ADR for power error
- [x] Update algorithm description

### CODE REVIEW

**Comment by dominicm00** [comment]
Jam calculates the source set size for piecemeal using an exponential backoff. This is rather cumbersome to implement in Ham's data structures, and is also very slow, so we calculate the exact length of the command ahead of time.

Doing so is slightly complex because of Jam's variable expansion. Namely, the source list could form a product with another variable (`$(X)-$(2) => x-1 y-1 x-2 y-2`), a product with itself (`$(2)-$(2) => 1-1 1-2 2-1 2-2`), or be nested within other variables (`X = 2 ; $((X)) => 1 2`).

The algorithm just pushed is able to incrementally calculate the cost of each new source with just two initial variable expansions, and relatively cheap combinatorics. This makes it very easy to simply iterate until the command is larger than the maximum limit, create a command, and repeat ad infinitum.

File: src/make/Processor.cpp [hunk 1]
```diff
@@ -944,80 +960,257 @@ Processor::_BuildCommand(data::RuleActionsCall* actionsCall)
...
 	// reinstate the old local variable scope and the built-in variables
 	fEvaluationContext.SetLocalScope(oldLocalScope);
 	fEvaluationContext.SetBuiltInVariables(oldBuiltInVariables);
-	return canceled ? nullptr
-					: new Command(actionsCall, commandLine, boundTargets);
+	return;
+}
+
+data::StringListList
+Processor::_PiecemealWords(
```
Comment (stippi) [inline src/make/Processor.cpp:1025]: Not sure if this fits Processor. Would move to some utility class.

Comment (dominicm00) [reply src/make/Processor.cpp:1039]: It's only useful in the context of _BuildCommands, so I'm not sure where else would be a good place to extract this.

**Comment by dominicm00** [comment]
While testing piecemeal on higher-power commands, I realized that the semantics Jam implies, and that I implemented, may not actually be desirable. 

Say you have the following situation:
```
# 2 = a b c ;
actions piecemeal maxline 20 EchoSources
{
    echo $(2)-$(2)
}
```

You would probably expect the result to look like this:
```
# Each echo command only has space for 4 pairs at a time
a-a a-b a-c b-a
b-b b-c c-a c-b
c-c
```

But it would actually look like this:
```
a-a a-b b-a b-b
c
```

Per the letter of the law, this is correct. The source set `{a, b, c}` was too large for the command buffer, so it was split into two separate commands running on the source sets `{a, b}` and `{c}`. However, this means you lose source pairs you would have had otherwise.

That said, there's no robust way to give the "obvious" behavior either. Say the action was this instead:
```
# 2 = a b c ;
actions piecemeal maxline 20 EchoSources
{
    echo $(2)-$(2) $(2)
}
```

How would you split this up? Splitting by the elements in the expanded word only works if there is exactly one word using the source variable.

The best behavior here is _probably_ to refuse piecemeal actions that have higher power words, as I don't see any practical reason for someone to have to do something like `$(2)-$(2)`. Higher multiplicities (e.g. `$(X)-$(2)`) can still be supported normally. The algorithm makes detecting higher power words trivial, and refusing them would also simplify the calculations.

**Pull request merged**